File: 132-palindrome-partitioning-ii/132-palindrome-partitioning-ii.py
# H | DP - Word Break and Fountain Type | T: O(n^2) | S: O(n^2)

class Solution:
    def minCut(self, s: str) -> int:
        def find_pal(i: int, j: int, t: dict, s: str) -> bool:
            if i >= j:
                return True
            if (i, j) in t:
                return t[(i, j)]
            if s[i]!=s[j]:
                t[(i, j)]= False
            else:
                t[(i, j)] = find_pal(i+1, j-1, t, s)
            return t[(i, j)]

        t = {}
        res = [2002 for i in range(len(s)+1)]
        res[0] = 0
        for i in range(len(s)):
            if res[i]==2002:
                continue
            for j in range(len(s)-1,i-1,-1):
                if find_pal(i,j,t,s):
                    res[j+1]=min(res[j+1],res[i]+1)
            
        return res[-1]-1

File: 132-palindrome-partitioning-ii/132-palindrome-partitioning-ii.cpp
class Solution {
public:
    int minCut(string s) {
        int n=s.length();
        vector<int>t(n);
        
        for(int i=0;i<n;i++) t[i]=i;
        
        for(int i=0;i<n;i++){
            
            for(int j=0;j<n && i-j>=0 && j+i<n;j++){
                if(s[i-j]==s[j+i]) {
                    if(i==j) t[i+j]=0;
                    else t[j+i]=min(t[j+i],t[i-j-1]+1);
                }
                else break;
            }
            
            for(int j=0;j<n && i-j>=0 && j+i+1<n;j++){
                if(s[i-j]==s[j+i+1]) {
                    if(i==j) t[i+j+1]=0;
                    else t[j+i+1]=min(t[j+i+1],t[i-j-1]+1);
                }
                else break;
            }
            
        }
        // for(int num:t) cout<<num<<endl;
        return t[n-1];
    }
};